readthedocs/telemetry/collectors.py

```diff
@@ -0,0 +1,261 @@
+"""Data collectors."""
+
+import json
+
+import structlog
+
+log = structlog.get_logger(__name__)
+
+
+class BuildDataCollector:
+
+    """
+    Build data collector.
+
+    Collect data from a runnig build.
+    """
+
+    def __init__(self, environment):
+        self.environment = environment
+        self.build = self.environment.build
+        self.project = self.environment.project
+        self.version = self.environment.version
+        self.config = self.environment.config
+
+        log.bind(
+            build_id=self.build["id"],
+            project_slug=self.project.slug,
+            version_slug=self.version.slug,
+        )
+
+    @staticmethod
+    def _safe_json_loads(content, default=None):
+        # pylint: disable=broad-except
+        try:
+            return json.loads(content)
+        except Exception:
+            log.info(
+                "Error while loading JSON content.",
+                exc_info=True,
+            )
+            return default
+
+    def run(self, *args, **kwargs):
+        build_cmd = self.environment.run(*args, record=False, **kwargs)
+        return build_cmd.exit_code, build_cmd.output, build_cmd.error
+
+    def collect(self):
+        """
+        Collect all relevant data from the runnig build.
+
+        Data that can be extracted from the database (project/organization)
+        isn't collected here.
+        """
+        data = {}
+        data["config"] = {"user": self.config.source_config}
```

Comment:
Aren't we saving the `final` here as well?

Reply:
we have that value on the build model https://github.com/readthedocs/readthedocs.org/blob/50c3f3f2b38430d4fd44d53643c1b84fd7eb6630/readthedocs/telemetry/models.py#L107-L107

Reply:
Will this `config.user` saved into the telemetry database be empty if the user doesn't have a `.readthedocs.yaml` or it will have the same problem we currently have that it contains the default values for that particular project making it impossible to filter by?

I'd like to get `config.user: null` if the user doesn't have a `.readthedocs.yaml` and `config.final: {...}` being the config file used by the build with the default values we auto-populated. This will allow us to answer the question we are trying to answer in https://github.com/readthedocs/meta/discussions/15#discussioncomment-2656380 cc @ericholscher

Reply:
config.user will be empty if the user didn't have a config file, config.final will have the defaults merged.

Reply:
Excellent!